eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -290,59 +292,63 @@ feesOK pp tx (UTxO utxo) = do
+type AlonzoUtxoNeeds era =
+  ( HasField "txrdmrs" (Core.Witnesses era) (Redeemers era),
     HasField "_keyDeposit" (Core.PParams era) Coin,
     HasField "_poolDeposit" (Core.PParams era) Coin,
     HasField "_maxTxSize" (Core.PParams era) Natural,
-    HasField "_prices" (Core.PParams era) Prices,
     HasField "_maxTxExUnits" (Core.PParams era) ExUnits,
     HasField "_coinsPerUTxOWord" (Core.PParams era) Coin,
     HasField "_maxValSize" (Core.PParams era) Natural,
-    HasField "_collateralPercentage" (Core.PParams era) Natural,
     HasField "_maxCollateralInputs" (Core.PParams era) Natural,
-    Core.Tx era ~ Alonzo.ValidatedTx era,
-    Core.Witnesses era ~ TxWitness era,
-    Era.TxSeq era ~ Alonzo.TxSeq era,
     HasField "vldt" (Core.TxBody era) ValidityInterval,
     HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
     HasField "collateral" (Core.TxBody era) (Set (TxIn (Crypto era))),
     HasField "mint" (Core.TxBody era) (Core.Value era),
     HasField "wdrls" (Core.TxBody era) (Wdrl (Crypto era)),
     HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
     HasField "datahash" (Core.TxOut era) (StrictMaybe (DataHash (Crypto era))),
-    ToCBOR (Core.Value era),
     HasField "txnetworkid" (Core.TxBody era) (StrictMaybe Network),
     HasField "address" (Core.TxOut era) (Addr (Crypto era))
+  )
+
+-- | The UTxO transition rule for the Alonzo eras.
+genericAlonzoUtxo ::
+  forall era utxo.
+  ( Era era,
+    -- STS properties
+    STS (utxo era),
+    Signal (utxo era) ~ ValidatedTx era,
+    Environment (utxo era) ~ Shelley.UtxoEnv era,
+    BaseM (utxo era) ~ ShelleyBase,
+    State (utxo era) ~ Shelley.UTxOState era,
+    PredicateFailure (utxo era) ~ UtxoPredicateFailure era,
+    -- instructions for calling UTXOS from utxo
+    Embed (Core.EraRule "UTXOS" era) (utxo era),
+    Environment (Core.EraRule "UTXOS" era) ~ Shelley.UtxoEnv era,
+    State (Core.EraRule "UTXOS" era) ~ Shelley.UTxOState era,
+    Signal (Core.EraRule "UTXOS" era) ~ Core.Tx era,
+    -- specific wierdness
+    Core.Tx era ~ ValidatedTx era,
+    ToCBOR (Core.Value era), -- we have to serialise Value to get a compact one.
+    -- substructural properties for (Era era)
+    AlonzoUtxoNeeds era
   ) =>
-  TransitionRule (AlonzoUTXO era)
-utxoTransition = do
+  UtxoDelta utxo era ->
+  TransitionRule (utxo era)
+genericAlonzoUtxo delta = do
   TRC (Shelley.UtxoEnv slot pp stakepools _genDelegs, u, tx) <- judgmentContext
   let Shelley.UTxOState utxo _deposits _fees _ppup _ = u
-
   {-   txb := txbody tx   -}
   {-   (,i_f) := txvldttx   -}
-  let txb = body tx
+  let txb = getField @"body" tx
       vi@(ValidityInterval _ i_f) = getField @"vldt" txb
-      inputsAndCollateral =
-        Set.union
-          (getField @"inputs" txb)
-          (getField @"collateral" txb)
+      allInputs =
+        Set.unions
+          [ getField @"inputs" txb,
+            getField @"collateral" txb,
+            referenceInputs' delta txb
```

Comment:
I don't think it is right to introduce concepts into past eras from the future eras. There was no reference inputs in Alonzo

Reply:
In order to generalize, we need to anticipate changes that will occur in future eras. Of course we can't always do that, but we can by adding these things in later, and presenting one consistent story across all eras.  Readers are not interested in the path we took to get to the current state, but in a consistent story about the current state. Many of our "differences" are adding new things, that are consistent with the past, as long as they have a zero or null value in past eras, that did not originally include them. Examples are the minting field, and the collateral inputs. We tell the story that we aways had these in mind, but did not bother to store them in Shelley, and Allegra (for mint) and for Shelley, Allegra, and Mary for (collateral) because the were aways, 0 for mint, and emptyset for collateral.